### PR TITLE
Add capibmadm tool related targets

### PR DESCRIPTION
**Pull request by Prajyot-Parab**
**What this PR does / why we need it**:
- Add `capibmadm` tool related targets
- Introduce version package

Output :
```
% ./capibmadm version -o short
v0.5.0-alpha.0.54-c4d887210d55c2-dirty

% ./capibmadm version -o yaml 
clientversion:
  major: "0"
  minor: "5"
  gitversion: v0.5.0-alpha.0.54-c4d887210d55c2-dirty
  gitcommit: c4d887210d55c27ff301def53f55b7ce044a4f4d
  gittreestate: dirty
  builddate: "2023-04-11T18:10:41Z"
  goversion: go1.19.6
  compiler: gc
  platform: darwin/amd64

% ./capibmadm version -o json
{
  "ibmcloudProviderVersion": {
    "major": "0",
    "minor": "5",
    "gitVersion": "v0.5.0-alpha.0.54-c4d887210d55c2-dirty",
    "gitCommit": "c4d887210d55c27ff301def53f55b7ce044a4f4d",
    "gitTreeState": "dirty",
    "buildDate": "2023-04-11T18:10:41Z",
    "goVersion": "go1.19.6",
    "compiler": "gc",
    "platform": "darwin/amd64"
  }
}
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
Add capibmadm tool related targets
```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -35,14 +41,20 @@ COPY pkg/ pkg/
 COPY util/ util/
 
 # Build
+ARG package=.
```
Comment (mkumatag) [inline Dockerfile:44]: I don't see this `package` ARG been used/passed anywhere!

File: Dockerfile [hunk 2]
```diff
@@ -35,14 +41,20 @@ COPY pkg/ pkg/
...
 ARG ARCH
-RUN CGO_ENABLED=0 GOOS=linux GOARCH=${ARCH} GO111MODULE=on go build -ldflags "-extldflags '-static'" -a -o manager main.go
+ARG LDFLAGS
```
Comment (mkumatag) [inline Dockerfile:46]: same for `LDFLAGS` as well

File: Dockerfile [hunk 3]
```diff
@@ -35,14 +41,20 @@ COPY pkg/ pkg/
...
+RUN --mount=type=cache,target=/root/.cache/go-build \
+    --mount=type=cache,target=/go/pkg/mod \
+    --mount=type=cache,target=/root/.local/share/golang \
+    CGO_ENABLED=0 GOOS=linux GOARCH=${ARCH} GO111MODULE=on go build -ldflags "${LDFLAGS} -extldflags '-static'"  -o manager ${package}
+ENTRYPOINT [ "/start.sh", "/workspace/manager" ]
```
Comment (mkumatag) [inline Dockerfile:51]: This is just an intermediate container, is setting `ENTRYPOINT` here will make any difference?

File: Makefile [hunk 4]
```diff
@@ -304,6 +319,31 @@ release-alias-tag: ## Add the release alias tag to the last build tag
...
+	RELEASE_BINARY=./cmd/capibmadm GOOS=linux GOARCH=ppc64le $(MAKE) release-binary
+	RELEASE_BINARY=./cmd/capibmadm GOOS=linux GOARCH=amd64 $(MAKE) release-binary
+	RELEASE_BINARY=./cmd/capibmadm GOOS=linux GOARCH=arm64 $(MAKE) release-binary
+	RELEASE_BINARY=./cmd/capibmadm GOOS=darwin GOARCH=amd64 $(MAKE) release-binary
+	RELEASE_BINARY=./cmd/capibmadm GOOS=darwin GOARCH=arm64 $(MAKE) release-binary
+	RELEASE_BINARY=./cmd/capibmadm GOOS=windows GOARCH=amd64 EXT=.exe $(MAKE) release-binary
+	RELEASE_BINARY=./cmd/capibmadm GOOS=windows GOARCH=arm64 EXT=.exe $(MAKE) release-binary
+
+.PHONY: release-binary
+release-binary: $(RELEASE_DIR) versions.mk build-toolchain ## Release binary
```
Comment (mkumatag) [inline Makefile:333]: why do we need this `versions.mk`?

File: Makefile [hunk 5]
```diff
@@ -304,6 +319,31 @@ release-alias-tag: ## Add the release alias tag to the last build tag
...
+	docker run \
+		--rm \
+		-e CGO_ENABLED=0 \
+		-e GOOS=$(GOOS) \
+		-e GOARCH=$(GOARCH) \
+		--mount=source=gocache,target=/go/pkg/mod \
+		--mount=source=gocache,target=/root/.cache/go-build \
+		-v "$$(pwd):/workspace$(DOCKER_VOL_OPTS)" \
```
Comment (mkumatag) [inline Makefile:341]: where is this `DOCKER_VOL_OPTS` variable defined?

File: Makefile [hunk 6]
```diff
@@ -304,6 +319,31 @@ release-alias-tag: ## Add the release alias tag to the last build tag
...
+		-w /workspace \
+		$(TOOLCHAIN_IMAGE) \
+		go build -ldflags '$(LDFLAGS) -extldflags "-static"' \
```
Comment (mkumatag) [inline Makefile:344]: Where is this `LDFLAGS` flag defined?

File: cmd/capibmadm/cmd/powervs/doc.go [hunk 7]
```diff
@@ -14,5 +14,5 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// Package powervs contains the commands to operate on Power VS resources.
+// Package powervs contains the commands to operate on PowerVS resources.
```
Comment (mkumatag) [inline cmd/capibmadm/cmd/powervs/doc.go:17]: lets address part of a different PR

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (Prajyot-Parab) [reply Dockerfile]: Fixed.

File: Dockerfile (hunk 2 above)
Comment (Prajyot-Parab) [reply Dockerfile:45]: Fixed.

File: Dockerfile (hunk 3 above)
Comment (Prajyot-Parab) [reply Dockerfile]: Removed not required.

File: Makefile (hunk 4 above)
Comment (Prajyot-Parab) [reply Makefile]: Removed not required.

File: Makefile (hunk 5 above)
Comment (Prajyot-Parab) [reply Makefile]: Removed not needed.

File: Makefile (hunk 6 above)
Comment (Prajyot-Parab) [reply Makefile:347]: Fixed, defined.

File: cmd/capibmadm/cmd/powervs/doc.go (hunk 7 above)
Comment (Prajyot-Parab) [reply cmd/capibmadm/cmd/powervs/doc.go]: Noted.

**Comment by mkumatag** [comment]
@k8s-ci-robot are you sure about the CLA?

File: Makefile [hunk 8]
```diff
@@ -55,11 +56,17 @@ PULL_BASE_REF ?= $(RELEASE_TAG) # PULL_BASE_REF will be provided by Prow
...
+# image name used to build the cmd/capibmadm
+TOOLCHAIN_IMAGE := toolchain
+
 TAG ?= dev
 ARCH ?= amd64
 ALL_ARCH ?= amd64 ppc64le arm64
 PULL_POLICY ?= Always
 
+# Set build time variables including version details
+LDFLAGS := $(shell ./hack/version.sh)
```
Comment (mkumatag) [inline Makefile:68]: This basically goes and sets the version in the version package but I don't see version is exist in the repository, so we might need to introduce to tag proper version with the tool we are shipping.

File: Makefile [hunk 9]
```diff
@@ -304,6 +322,31 @@ release-alias-tag: ## Add the release alias tag to the last build tag
 release-templates: $(RELEASE_DIR) ## Generate release templates
 	cp templates/cluster-template*.yaml $(RELEASE_DIR)/
 
+.PHONY: release-binaries
+release-binaries: ## Builds the binaries to publish with a release
```
Comment (mkumatag) [inline Makefile:326]: and also we need to call this target in the release target so that we generate this artefact as well when we release

Comment (mkumatag) [reply Makefile:326]: and we also need to explore how to ship it for the pre-GA builds

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: Makefile (hunk 9 above)
Comment (Prajyot-Parab) [reply Makefile:328]: Fixed.

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: cmd/capibmadm/cmd/version/version.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+	cmd := &cobra.Command{
+		Use:   "version",
+		Short: "Print version of capibmadm",
+		Args:  cobra.ExactArgs(0),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return RunVersion(out, cmd)
+		},
+	}
+
+	cmd.Flags().StringP("output", "o", "", "The output format of the result. Supported printer types: short, yaml, json")
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply cmd/capibmadm/cmd/version/version.go:48]: Not needed, fixed it to print minimalistic info.

File: cmd/capibmadm/cmd/version/version.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+	return cmd
+}
+
+// RunVersion provides the version information of capibmadm in format depending on arguments
+// specified in cobra.Command.
+func RunVersion(out io.Writer, cmd *cobra.Command) error {
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply cmd/capibmadm/cmd/version/version.go:55]: Fixed.

File: version/version.go [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+	Major                      string `json:"major,omitempty"`
+	Minor                      string `json:"minor,omitempty"`
+	GitVersion                 string `json:"gitVersion,omitempty"`
+	GitCommit                  string `json:"gitCommit,omitempty"`
+	GitTreeState               string `json:"gitTreeState,omitempty"`
+	BuildDate                  string `json:"buildDate,omitempty"`
+	GoVersion                  string `json:"goVersion,omitempty"`
+	VpcSdkVersion              string `json:"vpcSdkVersion,omitempty"`
+	PlatformServicesSdkVersion string `json:"platformServicesSdkVersion,omitempty"`
+	IbmSdkVersion              string `json:"ibmSdkVersion,omitempty"`
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply version/version.go]: The info may not be useful to the user, also we were not mapping those values just printing as part of tool version specifically. [removed those fields] - fixed.

**Comment by mkumatag** [comment]
@Prajyot-Parab can you paste the sample output of the version command in the description.?

**Comment by mkumatag** [comment]
```
% ./capibmadm-darwin-amd64 version
version.Info{Major:"0", Minor:"5", GitVersion:"v0.5.0-alpha.0.45-4171c845f378b6-dirty", GitCommit:"4171c845f378b6cd19c3b904a12ba6798083a5ff", GitTreeState:"dirty", BuildDate:"2023-04-05T18:12:15Z", GoVersion:"go1.20.2", Compiler:"gc", Platform:"darwin/amd64"}

```

this output is not a human readable format, I'm interested in printing more human readable format, lets add back the logic of json, yaml and short(default)

**Review by mkumatag — APPROVED** [review]
/lgtm

**Pull request merged**